File: 2/2-2.pas
// Не работает
function f(dt: array of boolean): boolean := 
  (dt[1] and (not dt[2])) or (dt[2] = dt[3]) or (not dt[0]);

begin
  foreach var vars in |False, True|.Cartesian(4) do
  begin
    var table := |(vars[0], vars[1], False, False), 
                  (True, False, vars[2], False), 
                  (True, False, True, vars[3])|;
    if table.Length = table.ToHashSet.Count then
    begin
      foreach var per in |0,1,2,3|.Permutations do
      begin
        var tr := (table[0].ToArray.Numerate.OrderBy(x -> per[x[0]]).Select(x -> x[1])).ToArray;
        if f(tr) = false then
          print();
      end;
    end;
  end;
end.

File: 2/2-2.py
from itertools import product, permutations


def f(w, x, y, z):
    return (x and (not y)) or (y == z) or (not w)


for a1, a2, a3, a4 in product([0, 1], repeat=4):
    table = [(a1, a2, 0, 0), (1, 0, a3, 0), (1, 0, 1, a4)]
    if len(table) == len(set(table)):
        for per in permutations('wxyz'):
            if [f(**dict(zip(per, row))) for row in table] == [0, 0, 0]:
                print(per)
